connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreMockitoTest.java

```diff
@@ -347,6 +380,165 @@ public void testPutConnectorConfigWithTargetState() throws Exception {
+                .when(configLog).readToEnd();
+        expectRead(CONNECTOR_TASK_COUNT_RECORD_KEYS.get(0), CONFIGS_SERIALIZED.get(0), CONNECTOR_TASK_COUNT_RECORD_STRUCTS.get(0));
+
+        // Should succeed now
+        configStorage.claimWritePrivileges();
+        configStorage.putTaskCountRecord(CONNECTOR_IDS.get(0), 6);
+
+        verify(fencableProducer).beginTransaction();
+        verify(fencableProducer).commitTransaction();
+
+        // Try to write a connector config
+        when(converter.fromConnectData(TOPIC, KafkaConfigBackingStore.CONNECTOR_CONFIGURATION_V0, CONNECTOR_CONFIG_STRUCTS.get(0)))
+                .thenReturn(CONFIGS_SERIALIZED.get(1));
+        // Get fenced out
+        doThrow(new ProducerFencedException("Better luck next time"))
+                .doNothing()
+                .when(fencableProducer).commitTransaction();
+
+        // Should fail again when we get fenced out
+        assertThrows(PrivilegedWriteException.class, () -> configStorage.putConnectorConfig(CONNECTOR_IDS.get(1), SAMPLE_CONFIGS.get(0), null));
+
+        verify(fencableProducer, times(2)).beginTransaction();
+        verify(fencableProducer).close(Duration.ZERO);
+
+        // Should fail if we retry without reclaiming write privileges
+        assertThrows(IllegalStateException.class, () -> configStorage.putConnectorConfig(CONNECTOR_IDS.get(1), SAMPLE_CONFIGS.get(0), null));
+
+        // In the meantime, write a target state (which doesn't require write privileges)
+        when(converter.fromConnectData(TOPIC, KafkaConfigBackingStore.TARGET_STATE_V1, TARGET_STATE_PAUSED))
+                .thenReturn(CONFIGS_SERIALIZED.get(1));
+        when(configLog.sendWithReceipt("target-state-" + CONNECTOR_IDS.get(1), CONFIGS_SERIALIZED.get(1)))
+                .thenReturn(producerFuture);
+        when(producerFuture.get(anyLong(), any(TimeUnit.class))).thenReturn(null);
+
+        // Should succeed even without write privileges (target states can be written by anyone)
+        configStorage.putTargetState(CONNECTOR_IDS.get(1), TargetState.PAUSED);
+
+        // Reclaim write privileges and successfully write the config
+        when(converter.toConnectData(TOPIC, CONFIGS_SERIALIZED.get(2)))
+                .thenReturn(new SchemaAndValue(null, structToMap(CONNECTOR_CONFIG_STRUCTS.get(0))));
+
+        // Should succeed if we re-claim write privileges
+        configStorage.claimWritePrivileges();
+        configStorage.putConnectorConfig(CONNECTOR_IDS.get(1), SAMPLE_CONFIGS.get(0), null);
+
+        verify(fencableProducer, times(3)).beginTransaction();
+        verify(fencableProducer, times(3)).commitTransaction();
+        verify(configUpdateListener).onConnectorConfigUpdate(CONNECTOR_IDS.get(1));
+
+        configStorage.stop();
+        verify(configLog).stop();
+        verify(fencableProducer, times(2)).close(Duration.ZERO);
```

Comment:
Could you add `verify(configStorage, times(2)).createFencableProducer();` to remind readers that we return the same `fencableProducer` instead of creating new one. Otherwise, `verify(fencableProducer, times(2)).close(Duration.ZERO);` is weird to me since it should not be closed two times in production.